cloudnew/ec2.go

```diff
@@ -0,0 +1,518 @@
+	}
+	grip.Info(message.Fields{
+		"msg":     "canceling spot request",
+		"spot_id": id,
+	})
+	if err = m.client.CancelSpotInstanceRequests(&ec2.CancelSpotInstanceRequestsInput{
+		SpotInstanceRequestIds: []*string{makeStringPtr(id)},
+	}); err != nil {
+		grip.Error(message.Fields{
+			"msg":     "failed to cancel spot request",
+			"spot_id": id,
+			"err":     err,
+		})
+		return false, errors.Wrapf(err, "Failed to cancel spot request for host %s", id)
+	}
+	if spotDetails == cloud.StatusRunning || spotDetails == cloud.StatusPending {
+		return true, nil
+	}
+	return false, nil
+}
+
+// IsUp returns whether a host is up.
+func (m *ec2Manager) IsUp(h *host.Host) (bool, error) {
+	if err := m.client.Create(m.credentials); err != nil {
+		return false, errors.Wrap(err, "error creating client")
+	}
+	info, err := m.getInstanceInfo(h.Id)
+	if err != nil {
+		return false, err
+	}
+	if ec2StatusToEvergreenStatus(*info.State.Name).String() == cloud.EC2StatusRunning {
+		return true, nil
+	}
+	return false, nil
+}
+
+// OnUp is called when the host is up.
+func (m *ec2Manager) OnUp(*host.Host) error {
+	// Unused since we set tags in SpawnHost()
+	return nil
+}
+
+// IsSSHReachable returns whether or not the host can be reached via SSH.
+func (m *ec2Manager) IsSSHReachable(h *host.Host, keyName string) (bool, error) {
+	opts, err := m.GetSSHOptions(h, keyName)
+	if err != nil {
+		return false, err
+	}
+	return hostutil.CheckSSHResponse(context.TODO(), h, opts)
+}
+
+// GetDNSName returns the DNS name for the host.
+func (m *ec2Manager) GetDNSName(h *host.Host) (string, error) {
+	info, err := m.getInstanceInfo(h.Id)
+	if err != nil {
+		return "", err
+	}
+	return *info.PublicDnsName, nil
+}
+
+// GetSSHOptions returns the command-line args to pass to SSH.
+func (m *ec2Manager) GetSSHOptions(h *host.Host, keyName string) ([]string, error) {
+	if keyName == "" {
+		return nil, errors.New("No key specified for EC2 host")
+	}
+	opts := []string{"-i", keyName}
+	hasKnownHostsFile := false
+
+	for _, opt := range h.Distro.SSHOptions {
+		opt = strings.Trim(opt, " \t")
+		opts = append(opts, "-o", opt)
+		if strings.HasPrefix(opt, "UserKnownHostsFile") {
+			hasKnownHostsFile = true
+		}
+	}
+
+	if !hasKnownHostsFile {
+		opts = append(opts, "-o", "UserKnownHostsFile=/dev/null")
+	}
+	return opts, nil
+}
+
+// TimeTilNextPayment returns how long until the next payment is due for a host.
+func (m *ec2Manager) TimeTilNextPayment(host *host.Host) time.Duration {
+	return timeTilNextEC2Payment(host)
+}
+
+// GetInstanceName returns the name of an instance.
+func (m *ec2Manager) GetInstanceName(d *distro.Distro) string {
+	return d.GenerateName()
+}
+
+// makeInt64Ptr is necessary because Go does not allow you to write `&int64(1)`.
+func makeInt64Ptr(i int64) *int64 {
+	return &i
+}
+
+// makeStringPtr is necessary because Go does not allow you to write `&"foo"`.
+func makeStringPtr(s string) *string {
+	return &s
+}
```

Comment:
move these functions to an ec2_utils? file? or maybe to the utils package? or maybe into the same file as the client rather than the manager? (or maybe it makes sense to make wrappers in our client wrapper to help minimize the need to make these conversions.